src/menus/EditMenus.cpp

```diff
@@ -73,65 +79,173 @@ bool DoPasteText(AudacityProject &project)
    return false;
 }
 
-// Return true if nothing selected, regardless of paste result.
-// If nothing was selected, create and paste into NEW tracks.
+void ForEachCopiedWaveTrack(const TrackList& src,
+                            const TrackList& dst,
+                            const std::function<void(const WaveTrack& waveTrack)>& f)
+{
+   if(dst.Selected().empty())
+   {
+      for(auto waveTrack : src.Any<const WaveTrack>())
+         f(*waveTrack);
+   }
+   else
+   {
+      const auto srcTrackRange = src.Any<const WaveTrack>();
+      const auto dstTrackRange = dst.Any<const WaveTrack>();
+      auto srcTrack = srcTrackRange.begin();
+      auto dstTrack = dstTrackRange.begin();
+      auto lastCopiedTrack = srcTrack;
+      while(dstTrack != dstTrackRange.end() && srcTrack != srcTrackRange.end())
+      {
+         if(!(*dstTrack)->GetSelected())
+         {
+            ++dstTrack;
+            continue;
+         }
+         
+         auto srcChannelCount = TrackList::Channels(*srcTrack).size();
+         auto dstChannelCount = TrackList::Channels(*dstTrack).size();
+                  
+         while(srcChannelCount > 0 && dstChannelCount > 0)
+         {
+            f(**srcTrack);
```

Comment:
Here and elsewhere that you call `f`, what proves that there is not a null dereference?  Track iterators are allowed to dereference an end iterator, getting a null, but then the second dereference might be unsafe.

The function would look better with a test for empty src and early exit.

Reply:
Not sure that understand your question. `dst` is allowed to be empty, `src` is expected to be not empty (and it can't have null tracks, right?). I don't see where null pointer dereference could happen in that case. I can comment about preconditions, but it's not a publicly available function which's purpose is to simplify implementation of another function.

Reply:
Right, to prove no null dereference happens, you must assume src is not an empty track list, and that is guaranteed at the call site.

It's all happening within one file, yes, so I won't be pedantic about stating preconditions.